### ISSUE TITLE
Codeception >= 3.1.1 and PHP 7.3 PHP error

### ISSUE THREAD

**Issue by angrybrad** [Codeception/Codeception#5735]
#### What are you trying to achieve?

Run unit tests on https://github.com/craftcms/cms using PHP 7.3 and Codeception > 3.1.0.

#### What do you get instead?

This error:

> strpos(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior

In GroupManager.php line 134.

You can see it on Travis here: https://travis-ci.com/craftcms/cms/builds/134460005

> Provide console output if related. Use `-vvv` mode for more details.

```bash
➜  cms git:(develop) ✗ ./vendor/bin/codecept run unit -vvv
Codeception PHP Testing Framework v3.1.1
Powered by PHPUnit 7.5.17 by Sebastian Bergmann and contributors.
Running with seed:


In GroupManager.php line 134:

  [yii\base\ErrorException (8192)]
  strpos(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior


Exception trace:
  at /home/vagrant/dev/cms/vendor/codeception/codeception/src/Codeception/Lib/GroupManager.php:134
 yii\base\ErrorHandler->handleError() at /home/vagrant/dev/cms/src/web/ErrorHandler.php:80
 craft\web\ErrorHandler->handleError() at n/a:n/a
 strpos() at /home/vagrant/dev/cms/vendor/codeception/codeception/src/Codeception/Lib/GroupManager.php:134
 Codeception\Lib\GroupManager->groupsForTest() at /home/vagrant/dev/cms/vendor/codeception/codeception/src/Codeception/SuiteManager.php:124
 Codeception\SuiteManager->addToSuite() at /home/vagrant/dev/cms/vendor/codeception/codeception/src/Codeception/SuiteManager.php:113
 Codeception\SuiteManager->addToSuite() at /home/vagrant/dev/cms/vendor/codeception/codeception/src/Codeception/SuiteManager.php:102
 Codeception\SuiteManager->loadTests() at /home/vagrant/dev/cms/vendor/codeception/codeception/src/Codeception/Codecept.php:191
 Codeception\Codecept->runSuite() at /home/vagrant/dev/cms/vendor/codeception/codeception/src/Codeception/Codecept.php:160
 Codeception\Codecept->run() at /home/vagrant/dev/cms/vendor/codeception/codeception/src/Codeception/Command/Run.php:502
 Codeception\Command\Run->runSuites() at /home/vagrant/dev/cms/vendor/codeception/codeception/src/Codeception/Command/Run.php:397
 Codeception\Command\Run->execute() at /home/vagrant/dev/cms/vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /home/vagrant/dev/cms/vendor/symfony/console/Application.php:934
 Symfony\Component\Console\Application->doRunCommand() at /home/vagrant/dev/cms/vendor/symfony/console/Application.php:273
 Symfony\Component\Console\Application->doRun() at /home/vagrant/dev/cms/vendor/symfony/console/Application.php:149
 Symfony\Component\Console\Application->run() at /home/vagrant/dev/cms/vendor/codeception/codeception/src/Codeception/Application.php:107
 Codeception\Application->run() at /home/vagrant/dev/cms/vendor/codeception/codeception/codecept:43
```

### Details

Note that PHP 7.3 works fine on Codeception version < 3.1.1 and PHP 7.2 and 7.1 work fine work fine on Codception >= 3.1.1.

* Codeception version: 3.1.1 or greater
* PHP Version: 7.3.9
* Operating System: macOS 10.15.1
* Installation type: Composer
* List of installed packages (`composer show`)
```
behat/gherkin                      v4.6.0     Gherkin DSL parser for PHP 5.3
cebe/markdown                      1.2.1      A super fast, highly extensible markdown parser for PHP
codeception/codeception            3.1.1      BDD-style testing framework
codeception/mockery-module         0.3.0      Mockery Module for Codeception
codeception/phpunit-wrapper        7.7.1      PHPUnit classes used by Codeception
codeception/specify                0.4.6      BDD code blocks for PHPUnit and Codeception
codeception/stub                   3.0.0      Flexible Stub wrapper for PHPUnit's Mock Builder
codeception/verify                 1.2.0      BDD assertion library for PHPUnit
composer/ca-bundle                 1.2.4      Lets you find a path to the system CA bundle, and includes a fallback to the Mozilla CA bundle.
composer/composer                  1.6.3      Composer helps you declare, manage and install dependencies of PHP projects, ensuring you have the right stack everywhere.
composer/semver                    1.5.0      Semver library that offers utilities, version constraint parsing and validation.
composer/spdx-licenses             1.5.2      SPDX licenses list and validation library.
craftcms/oauth2-craftid            1.0.0.1    Craft OAuth 2.0 Client Provider for The PHP League OAuth2-Client
craftcms/plugin-installer          1.5.4      Craft CMS Plugin Installer
craftcms/server-check              1.1.7      Craft CMS Server Check
creocoder/yii2-nested-sets         0.9.0      The nested sets behavior for the Yii framework
doctrine/instantiator              1.2.0      A small, lightweight utility to instantiate objects in PHP without invoking their constructors
doctrine/lexer                     1.0.2      PHP Doctrine Lexer parser library that can be used in Top-Down, Recursive Descent Parsers.
egulias/email-validator            2.1.11     A library for validating emails against several RFCs
elvanto/litemoji                   1.4.4      A PHP library simplifying the conversion of unicode, HTML and shortcode emoji.
enshrined/svg-sanitize             0.9.2      An SVG sanitizer for PHP
ezyang/htmlpurifier                v4.12.0    Standards compliant HTML filter written in PHP
facebook/webdriver                 1.7.1      A PHP client for Selenium WebDriver
fzaninotto/faker                   v1.8.0     Faker is a PHP library that generates fake data for you.
guzzlehttp/guzzle                  6.4.1      Guzzle is a PHP HTTP client library
guzzlehttp/promises                v1.3.1     Guzzle promises library
guzzlehttp/psr7                    1.6.1      PSR-7 message implementation that also provides common utility methods
hamcrest/hamcrest-php              v2.0.0     This is the PHP port of Hamcrest Matchers
hoa/consistency                    1.17.05.02 The Hoa\Consistency library.
hoa/console                        3.17.05.02 The Hoa\Console library.
hoa/event                          1.17.01.13 The Hoa\Event library.
hoa/exception                      1.17.01.16 The Hoa\Exception library.
hoa/file                           1.17.07.11 The Hoa\File library.
hoa/iterator                       2.17.01.10 The Hoa\Iterator library.
hoa/protocol                       1.17.01.14 The Hoa\Protocol library.
hoa/stream                         1.17.02.21 The Hoa\Stream library.
hoa/ustring                        4.17.01.16 The Hoa\Ustring library.
justinrainbow/json-schema          5.2.9      A library to validate a json schema.
league/factory-muffin              v3.0.1     The goal of this package is to enable the rapid creation of objects for the purpose of testing.
league/flysystem                   1.0.57     Filesystem abstraction: Many filesystems, one API.
league/oauth2-client               2.4.1      OAuth 2.0 Client Library
mikehaertl/php-shellcommand        1.6.0      An object oriented interface to shell commands
mockery/mockery                    1.2.4      Mockery is a simple yet flexible PHP mock object framework
myclabs/deep-copy                  1.9.3      Create deep copies (clones) of your objects
opis/closure                       3.4.1      A library that can be used to serialize closures (anonymous functions) and arbitrary objects.
paragonie/random_compat            v9.99.99   PHP 5.x polyfill for random_bytes() and random_int() from PHP 7
phar-io/manifest                   1.0.3      Component for reading phar.io manifest information from a PHP Archive (PHAR)
phar-io/version                    2.0.1      Library for handling version information and constraints
phpdocumentor/reflection-common    2.0.0      Common reflection classes used by phpdocumentor to reflect the code structure
phpdocumentor/reflection-docblock  4.3.2      With this component, a library can provide support for annotations via DocBlocks or otherwise retrieve information that is embedded in a DocBlock.
phpdocumentor/type-resolver        1.0.1      A PSR-5 based resolver of Class names, Types and Structural Element Names
phpoption/phpoption                1.5.0      Option Type for PHP
phpspec/prophecy                   1.9.0      Highly opinionated mocking framework for PHP 5.3+
phpunit/php-code-coverage          6.1.4      Library that provides collection, processing, and rendering functionality for PHP code coverage information.
phpunit/php-file-iterator          2.0.2      FilterIterator implementation that filters files based on a list of suffixes.
phpunit/php-text-template          1.2.1      Simple template engine.
phpunit/php-timer                  2.1.2      Utility class for timing
phpunit/php-token-stream           3.1.1      Wrapper around PHP's tokenizer extension.
phpunit/phpunit                    7.5.17     The PHP Unit Testing framework.
pixelandtonic/imagine              1.2.2.1    Image processing for PHP 5.3
psr/container                      1.0.0      Common Container Interface (PHP FIG PSR-11)
psr/http-message                   1.0.1      Common interface for HTTP messages
psr/log                            1.1.1      Common interface for logging libraries
ralouphie/getallheaders            3.0.3      A polyfill for getallheaders.
sebastian/code-unit-reverse-lookup 1.0.1      Looks up which function or method a line of code belongs to
sebastian/comparator               3.0.2      Provides the functionality to compare PHP values for equality
sebastian/diff                     3.0.2      Diff implementation
sebastian/environment              4.2.2      Provides functionality to handle HHVM/PHP environments
sebastian/exporter                 3.1.2      Provides the functionality to export PHP variables for visualization
sebastian/global-state             2.0.0      Snapshotting of global state
sebastian/object-enumerator        3.0.3      Traverses array structures and object graphs to enumerate all referenced objects
sebastian/object-reflector         1.1.1      Allows reflection of object attributes, including inherited and non-public ones
sebastian/recursion-context        3.0.0      Provides functionality to recursively process PHP variables
sebastian/resource-operations      2.0.1      Provides a list of PHP built-in functions that operate on resources
sebastian/version                  2.0.1      Library that helps with managing the version number of Git-hosted PHP projects
seld/cli-prompt                    1.0.3      Allows you to prompt for user input on the command line, and optionally hide the characters they type
seld/jsonlint                      1.7.2      JSON Linter
seld/phar-utils                    1.0.1      PHAR file format utilities, for when PHP phars you up
swiftmailer/swiftmailer            v6.2.1     Swiftmailer, free feature-rich PHP mailer
symfony/browser-kit                v4.3.5     Symfony BrowserKit Component
symfony/console                    v4.3.5     Symfony Console Component
symfony/css-selector               v4.3.5     Symfony CssSelector Component
symfony/dom-crawler                v4.3.5     Symfony DomCrawler Component
symfony/event-dispatcher           v4.3.5     Symfony EventDispatcher Component
symfony/event-dispatcher-contracts v1.1.7     Generic abstractions related to dispatching event
symfony/filesystem                 v4.3.5     Symfony Filesystem Component
symfony/finder                     v4.3.5     Symfony Finder Component
symfony/polyfill-ctype             v1.12.0    Symfony polyfill for ctype functions
symfony/polyfill-iconv             v1.12.0    Symfony polyfill for the Iconv extension
symfony/polyfill-intl-grapheme     v1.12.0    Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-idn          v1.12.0    Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer   v1.12.0    Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring          v1.12.0    Symfony polyfill for the Mbstring extension
symfony/polyfill-php72             v1.12.0    Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/polyfill-php73             v1.12.0    Symfony polyfill backporting some PHP 7.3+ features to lower PHP versions
symfony/process                    v4.3.5     Symfony Process Component
symfony/service-contracts          v1.1.7     Generic abstractions related to writing services
symfony/yaml                       v4.3.5     Symfony Yaml Component
theseer/tokenizer                  1.1.3      A small library for converting tokenized PHP source code into XML and potentially other formats
true/punycode                      v2.1.1     A Bootstring encoding of Unicode for Internationalized Domain Names in Applications (IDNA)
twig/twig                          v2.12.1    Twig, the flexible, fast, and secure template language for PHP
vlucas/phpdotenv                   v3.6.0     Loads environment variables from `.env` to `getenv()`, `$_ENV` and `$_SERVER` automagically.
voku/anti-xss                      4.1.15     anti xss-library
voku/arrayy                        5.14.2     Array manipulation library for PHP, called Arrayy!
voku/email-check                   3.0.2      email-check (syntax, dns, trash, ...) library
voku/portable-ascii                1.3.4      Portable ASCII library - performance optimized (ascii) string functions for php.
voku/portable-utf8                 5.4.25     Portable UTF-8 library - performance optimized (unicode) string functions for php.
voku/stop-words                    2.0.1      Stop-Words via PHP
voku/stringy                       5.1.1      A string manipulation library with multibyte support
voku/urlify                        4.1.1      PHP port of URLify.js from the Django project. Transliterates non-ascii characters for use in URLs.
webmozart/assert                   1.5.0      Assertions to validate method input/output with nice error messages.
webonyx/graphql-php                v0.12.6    A PHP port of GraphQL reference implementation
yii2tech/ar-softdelete             1.0.4      Provides support for ActiveRecord soft delete in Yii2
yiisoft/yii2                       2.0.21     Yii PHP Framework Version 2
yiisoft/yii2-composer              2.0.8      The composer plugin for Yii extension installer
yiisoft/yii2-debug                 2.1.10     The debugger extension for the Yii framework
yiisoft/yii2-queue                 2.1.0      Yii2 Queue Extension which supported DB, Redis, RabbitMQ, Beanstalk, SQS and Gearman
yiisoft/yii2-swiftmailer           2.1.2      The SwiftMailer integration for the Yii framework
zendframework/zend-escaper         2.6.1      Securely and safely escape HTML, HTML attributes, JavaScript, CSS, and URLs
zendframework/zend-feed            2.12.0     provides functionality for consuming RSS and Atom feeds
zendframework/zend-stdlib          3.2.1      SPL extensions, array utilities, error handlers, and more
```

**Comment by Naktibalda** [comment]
I guess that realpath returns FALSE in some cases in this line: https://github.com/Codeception/Codeception/blob/3.1.1/src/Codeception/Lib/GroupManager.php#L70
```php
$filePath = realpath(Configuration::projectDir() . $file);
```

Before last change that code simply did:
```php
$this->testsInGroups[$group][] = Configuration::projectDir() . $file;
```

@maks-rafalko Could you take a look at this issue too?

**Comment by bensquire** [comment]
I can also confirm this error. Had to restrict our codeception version to 3.1.0.

This might become especially problematic as people look to upgrade from PHP7.1

**Comment by Naktibalda** [comment]
@bensquire it is a very special issue, because majority of users is not affected.
Could you do some debugging and add something like 
```php
if ($filePath === false) {
  var_dump($file);
  die();
}
```
below line 70 of GroupManager that I linked above?
What value do you get when you run Codeception?

**Comment by bensquire** [comment]
Okay, thanks to your prompt I've worked out what the issue was for us. In our codeception.yml, one of our groups was misconfigured.

```
groups:
  unit: [test/codecept/unitDB, test/codecept/unit, test/codeception/integration]
```
instead of
```
groups:
  unit: [test/codecept/unitDB, test/codecept/unit, test/codecept/integration]
```

There could be an argument here for more robust validation/error handling?

Either way, thanks for the help! Hopefully @angrybrad has something similar going on?

**Comment by Naktibalda** [comment]
@bensquire Thanks, that's what I expected.
Would it be better to throw an exception that the file doesn't exist or to ignore it silently?

**Comment by bensquire** [comment]
@Naktibalda Personally I'd rather know I've misconfigured something so that I can fix it.

Thanks to you and the team btw for all the hard work.

**Comment by angrybrad** [comment]
@Naktibalda @bensquire Can confirm it was a bad path in my codeception.yml group.  🤦‍♂️

I'd probably prefer an explicit exception be thrown in that case as well.

Thanks for the debugging help!

**Comment by Naktibalda** [comment]
The patch was merged to 4.0 branch and it will be released as 4.0.0 really soon.
